source/coding/git.rst

```diff
@@ -1,12 +1,150 @@
+       :scale: 40 %
+       :align: right
+
+Setting up everything the first time
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. code-block:: bash
+
+    # Cloning the code
+    git clone $link_to_myFork
+    cd myFork
+
+    # Telling git who are the upstream and origin
+    git remote add upstream $link_to_upstream
+    git remote add origin $link_to_myFork # Should be set automatically
+    git remote -v # To verify everything.
+
+    # Creating branches
+    git branch myBranch1 # create it
+    git checkout myBranch1 # switch on it
+    git checkout -b myBranch2 # create AND switch
+    # Note that here, myBranch2 would be a copy of myBranch1.
+    # If instead you want to start from the master, switch to your master branch first.
+
+    # Delete a branch:
+    git branch -d myBranch # For a local branch
+    git push origin --delete myBranch # For a branch on origin
+
+Downloading updates
+^^^^^^^^^^^^^^^^^^^
+
+.. code-block:: bash
+
+    # From the origin to both your local repo and your workspace:
+    git pull origin master # master or any branch you want to update.
+                           # Note that this is equivalent to git fetch + git merge.
+
+    # From the upstream to your local repo
+    git pull upstream master
+
+    # From the upstream to your forked repo
+    # There is no way to do this direclty. You can update your local repo and
+    # send the update back up to your forked repo
+    git push origin master
+
+    # Updating one branch on your computer with the updates from the master:
+    # Update master. Then:
+    git rebase master myBranch
+
+
+Uploading your changes
+^^^^^^^^^^^^^^^^^^^^^^
+
+.. image:: /images/git5.jpeg
+   :scale: 80 %
+   :align: right
+
+Each update is called a **commit**. See figure [4]_.
+
+.. code-block:: bash
+
+    # From the workspace to your local repo:
+    git status # To see which files have been modified without telling git
+    git add myFile  # To add or update a file in git's memory.
+    git add -u      # To add all files that have been modified, but that git already knew
+    git status      # If you check again, everything in green will be recorded with the next command:
+    git commit -m "Message to explain the work I have been doing in this update"
+
+    # From the local repo to your forked repo
+    git push origin myBranch  # Will automatically send it to the same branch name in your forked repo.
+
+    # From your forked repo to the upstream repo:
+    # Use the Pull Request button on internet
+
+    # From your local repo to the upstream repo:
+    # DON'T DO THAT. Didn't you see that there is no such arrow on the figure!
+
+
+.. image:: /images/git4.png
+   :scale: 75 %
+   :align: right
+
+Merging branches
+^^^^^^^^^^^^^^^^
+
+Figure: [5]_.
+
+.. code-block:: bash
+
+    # Technique 1: merge. To merge branch2 (pink) into branch1 (yellow):
+    git checkout branch1
+    git merge branch2
+
+    # Technique 2: rebase
+    git rebase branch1 branch2
```

Comment:
Here, it is required that we specify a rebase on a branch that has already been published to a remote must be update to the remote with a git push --force